### PR TITLE
feature: support lxcfs

### LINKED ISSUE #140: [feature request] support lxcfs in pouch to improve isolation
**Issue Description**

[LXCFS](https://github.com/lxc/lxcfs) is a small FUSE filesystem written with the intention of making Linux containers feel more like a virtual machine.

Pouch takes lxcfs support as a high priority.

We need to add these things in pouch with lxcfs:

- [ ] Design document of pouch and lxcfs;
- [ ] API integration in Pouch via lxcfs;
- [ ] daemon side implementation
- [ ] test for lxcfs
  - [ ] unit test
  - [ ] integration test

/cc @CodeJuan

### PR DESCRIPTION
**Pull request by CodeJuan**
**1.Describe what this PR did**
Add lxcfs to make Pouch feel more like a real independent system

**2.Does this pull request fix one issue?** 

fixed #140 

**3.Describe how you did it**
1. Add lxcfs flag to pouchd, and start lxcfs process by pouchd
1. Add `EnableLxcfs` to struct ContainerConfig
1. If `EnableLxcfs` of ContainerCreateConfig is true, then bind lxcfs proc files to container(https://github.com/alibaba/pouch/issues/140#issuecomment-347425189)

**4.Describe how to verify it**
1. install and stop lxcfs
```
apt-get install lxcfs
service lxcfs stop
```
2. start pouchd with lxcfs
```
pouchd --enable-lxcfs=true --lxcfs=/usr/bin/lxcfs
```
3. create a container with lxcfs
```
curl --unix-socket  /var/run/pouchd.sock -H "Content-Type: application/json"   -d '{"Image": "docker.io/library/nginx:latest", "EnableLxcfs": true}' -X POST http://localhost/containers/create
```
4. start the container
```
pouch start [container ID]
```
5. check the uptime of container, the uptime should equal to start time of container
```
nsenter --target [pid of container] --mount --uts --ipc --net --pid -- cat /proc/uptime
```
**5.Special notes for reviews**

### CODE REVIEW

**Comment by allencloud** [comment]
Fantastic! 👍 🍻  @CodeJuan 
Could you add more details about Decribe what you did, how to verify this in PR description and so on?

File: main.go [hunk 1]
```diff
@@ -146,3 +155,42 @@ func initLog() {
...
+
+	lxcfsProcess := &exec.Process{
+		Path: cfg.LxcfsBinPath,
+		Args: []string{
+			cfg.LxcfsHome,
+		},
+	}
+	processes = append(processes, lxcfsProcess)
+
+	if cfg.LxcfsHome[len(cfg.LxcfsHome)-1:] == "/" {
```
Comment (skoo87) [inline main.go]: you can use "strings.HasSuffix(cfg.LxcfsHome, "/")"

**Comment by allencloud** [comment]
I tested on my local machine, and I found that in daemon's log:
```
INFO[2018-01-05 09:57:58.029216982] close containerio backend: raw-file, id: babd75467bec7a8764e7931f08165cad2fab42ef6b0820a20ff15d22e4c745d4
INFO[2018-01-05 09:57:58.029426883] close containerio backend: raw-file, id: babd75467bec7a8764e7931f08165cad2fab42ef6b0820a20ff15d22e4c745d4
INFO[2018-01-05 09:57:58.032138520] finished to subscribe io, backend: raw-file, id: babd75467bec7a8764e7931f08165cad2fab42ef6b0820a20ff15d22e4c745d4
INFO[2018-01-05 09:57:58.072541138] start to listen to: unix:///var/run/pouchd.sock
hierarchies:
  0: fd:   5: memory
  1: fd:   9: freezer
  2: fd:  10: blkio
  3: fd:  11: perf_event
  4: fd:  12: cpuset
  5: fd:  13: pids
  6: fd:  14: cpu,cpuacct
  7: fd:  15: hugetlb
  8: fd:  16: net_cls,net_prio
  9: fd:  17: devices
 10: fd:  18: name=systemd
fuse: bad mount point `/var/lib/lxc/lxcfs': No such file or directory
ERRO[2018-01-05 09:57:58.854817718] failed to wait process: /usr/bin/lxcfs, exit status 1
WARN[2018-01-05 09:57:58.854857402] process: /usr/bin/lxcfs exited
hierarchies:
  0: fd:   5: memory
  1: fd:   9: freezer
  2: fd:  10: blkio
  3: fd:  11: perf_event
  4: fd:  12: cpuset
  5: fd:  13: pids
  6: fd:  14: cpu,cpuacct
  7: fd:  15: hugetlb
  8: fd:  16: net_cls,net_prio
  9: fd:  17: devices
 10: fd:  18: name=systemd
fuse: bad mount point `/var/lib/lxc/lxcfs': No such file or directory
ERRO[2018-01-05 09:57:59.852909489] failed to wait process: /usr/bin/lxcfs, exit status 1
WARN[2018-01-05 09:57:59.853117317] process: /usr/bin/lxcfs exited
hierarchies:
  0: fd:   5: memory
  1: fd:   9: freezer
  2: fd:  10: blkio
  3: fd:  11: perf_event
```

I think we need to create this default directory. @CodeJuan 
After I created this dir mannually, error never happens.

File: main.go [hunk 2]
```diff
@@ -146,3 +155,42 @@ func initLog() {
...
+		cfg.LxcfsHome = cfg.LxcfsHome[0 : len(cfg.LxcfsHome)-2]
```
Comment (skoo87) [inline main.go]: What does the mean ???

File: main.go [hunk 3]
```diff
@@ -146,3 +155,42 @@ func initLog() {
...
 	logrus.SetFormatter(formatter)
 }
+
+// define lxcfs processe.
+func setLxcfsProcess(processes exec.Processes) exec.Processes {
+	if !cfg.IsLxcfsEnabled {
+		return processes
+	}
+
+	lxcfsProcess := &exec.Process{
```
Comment (skoo87) [inline main.go]: s/lxcfsProcess/process ?

**Comment by allencloud** [comment]
I have ran lxcfs, and I found one issue: if I execute `uptime`, the time is also from the host; if I execute `cat /proc/uptime`, it is the container's uptime.
```
# uptime
02:11:24 up 18 min,  0 users,  load average: 0.00, 0.00, 0.00
/ # cat /proc/uptime
 501.0 501.0
/ #
```

Is it reasonable? @CodeJuan

File: main.go (hunk 1 above)
Comment (CodeJuan) [reply main.go]: Remove the last char '/' to let path.Dir get the parent directory of lxcfs home

File: main.go (hunk 2 above)
Comment (CodeJuan) [reply main.go]: Remove the last char '/' to let `path.Dir` get the parent directory of lxcfs home

File: main.go (hunk 1 above)
Comment (CodeJuan) [reply main.go]: Yep, it's more readable.

File: main.go (hunk 2 above)
Comment (skoo87) [reply main.go]: You can use "Trim" .

File: apis/swagger.yml [hunk 4]
```diff
@@ -925,6 +925,10 @@ definitions:
         type: "array"
         items:
           type: "string"
+      EnableLxcfs:
```
Comment (allencloud) [inline apis/swagger.yml]: I think we need to add this parameter in API change to show the difference. Just record this and do this in the future.

**Comment by allencloud** [comment]
Please rebase this PR to master, since #511 has been merged and memory limit is supported. @CodeJuan

**Comment by CodeJuan** [comment]
@skoo87 @allencloud 
Thanks for coaching me, I have rebased the code to address the comments.
Please help to review. Thank you.

**Comment by Letty5411** [comment]
I have open a issue #522 to add lxcfs test support in CI after this PR is merged.

**Comment by allencloud** [comment]
LGTM

**Pull request merged**